docs/04-storage-overview.md

```diff
@@ -0,0 +1,327 @@
+
+Write is initiated by [`InsertExecutor`](https://github.com/risinglightdb/risinglight/blob/main/src/executor/insert.rs) or [`DeleteExecutor`](https://github.com/risinglightdb/risinglight/blob/main/src/executor/delete.rs) (and in the future, `UpdateExecutor`).
+
+The executor will start a write transaction, then appending data chunks into the transaction. Inside transaction structure, RisingLight will maintain a memtable (in BTree or just a vector based on whether the data needs to be sorted) during the transaction. When executors commit a transaction, memtable will be encoded into RowSets, written to disk, and added into the manifest.
+
+The public interface (like `Transaction`, `Table`) is defined in [`storage/mod.rs`](https://github.com/risinglightdb/risinglight/blob/main/src/storage/mod.rs). Simple as the interface looks like, there are a lot of things going on internally in *secondary*.
+
+The data chunk from `InsertExecutor` will be piped into `RowsetBuilder`, where data from each column will be fed into `ColumnBuilder`. `ColumnBuilder` will decide how many items a block contains, and pipe the data into `BlockBuilder`. `Block` is the minimum managing unit of *secondary*. Each block will occupy an entry in block cache. `ColumnBuilder` will also build a sparse index for the column, which contains the begin offset of each block in the `.col` file, and the first key of each block (if it is a sorted column).
+
+## Read Path
+
+[`TableScanExecutor`](https://github.com/risinglightdb/risinglight/blob/main/src/executor/table_scan.rs) will read from the storage. It will start a read transaction, and get data chunks from the storage.
+
+Also there are a lot of things going on internally in *secondary*. Upon starting the read transaction, *secondary* will take a snapshot of the on-disk merge tree structure by pinning the latest storage *version*. If a version is pinned, the files in that version won't be deleted (or more professionally speaking, vacuumed), and executors can read from those files safely.
+
+A table will typically contain multiple RowSets. *Secondary* will create an iterator over multiple RowSets either by using `MergeIterator` (to do a merge sort), or by using `ConcatIterator` (by yielding data one RowSet by RowSet). Under those iterators, there is one of the most fundamental and important iterator, `RowSetIterator`. `RowSetIterator` scans the underlying columns by using `ColumnIterator`, which uses `BlockIterator` internally. It will also take delete vectors into consideration, so as to filter out deleted rows.
+
+`RowSetIterator` also supports filter scan. Users can provide a filter expression to `RowSetIterator`, and the iterator will skip reading blocks to reduce I/O.
+
+There is a special column called `RowHandler` in *secondary*. This column is of int64 type, which contains RowSet id on upper 32 bits, and row offset in lower 32 bits. For example, if there is a table, which contains a RowSet with the following data:
```

Comment:
Should `RowHandler` be renamed to `RowHandle`? Because **handler** sounds like a function that processes sth, and **handle** usually means a reference (or unique ID) to some resource.

Reply:
That’s true! We can do a rename in both code and docs next time.